qe/drivers.py

```diff
@@ -266,6 +267,296 @@ def is_connected(det_i: Determinant, det_j: Determinant) -> Tuple[int, int]:
+                (set(det_a) | set(constraint)) - (set(det_a) & set(constraint))
+            ) & set(
+                constraint
+            )  # This set will have length 2; unpack
+            paa.append((a_unocc_1, a_unocc_2))
+
+        elif len(constraint_orbitals_occupied) == 2:
+            # (Two constraint orbitals occupied) e.g., a1, a2 \in |D_a> ->
+            #   A same-spin (aa) double (x_a, y_a) \in haa to (z_a, a_unocc), where z_a\not\in|D_a>, and z_a < a1 (if excited to z_a > a1, constraint ruined!)
+            (a_unocc,) = ((set(det_a) | set(constraint)) - (set(det_a) & set(constraint))) & set(
+                constraint
+            )  # The 1 unoccupied constraint orbital
+            det_unocc_a_orbs = self.all_orbs - set(det_a)  # Unocc orbitals in |D_a>
+            for z_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                # z < a_unocc trivially, no need to check they are distinct
+                paa.append((z_a, a_unocc))
+            #   No same spin (bb) excitations will satisfy constraint
+            #   An oppopsite spin (ab) double (x_a, y_b) \in \hab to (a_unocc, z_b), where z\not\in|D_b>
+            det_unocc_b_orbs = set(range(self.n_orb)) - set(det_b)  # Unocc orbitals in |D_b>
+            for _, z_b in enumerate(det_unocc_b_orbs):
+                pab.append((a_unocc, z_b))
+
+        elif len(constraint_orbitals_occupied) == 3:
+            # a1, a2, a3 \in |D_a> -> |D> satisfies constraint! ->
+            #   Any same-spin (aa) double (x_a, y_a) \in haa to (w_a, z_a), where w_a, z_a < a1
+            det_unocc_a_orbs = self.all_orbs - set(det_a)
+            for w_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                for z_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                    if w_a != z_a:
+                        paa.append((w_a, z_a))
+            #   Any same-spin (bb) double (x_b, y_b) \in hbb to (w_b, z_b)
+            det_unocc_b_orbs = self.all_orbs - set(det_b)  # Unocc orbitals in |D_a>
+            for _, w_b in enumerate(det_unocc_b_orbs):
+                for _, z_b in enumerate(det_unocc_b_orbs):
+                    if w_b != z_b:
+                        pbb.append((w_b, z_b))
+            #   Any oppospite-spin (ab) double (x_a, y_b) \in hab to (w_a, z_b), where w_a < a1
+            for w_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                for _, z_b in enumerate(det_unocc_b_orbs):
+                    pab.append((w_a, z_b))
+
+        # Create list of possible `holes` s.to constraint
+        if len(nonconstrained_orbitals_occupied) == 2:
+            # x_a, y_a \in |D_a> with x_a, y_a > a1 and \not\in {a1, a2, a3} -> A same-spin (aa) double (x_a, y_a) to (w_a, z_a) \in paa is necessary
+            # No same-spin (bb) or opposite-spin (ab) excitations will satisfy constraint
+            # New holes -> x, y
+            x_a, y_a = nonconstrained_orbitals_occupied  # This set will have length 2; unpack
```

Comment:
Do you have some symetry? how do you know the first one is `x_a` and the second is `y_a` or it doesn't matter?

Reply:
It doesn't matter, just passing particle hole pairs as sets